lib/puppet.rb

```diff
@@ -237,7 +237,7 @@ def self.base_context(settings)
       :ssl_context => proc { Puppet.runtime[:http].default_ssl_context },
       :http_session => proc { Puppet.runtime[:http].create_session },
       :plugins => proc { Puppet::Plugins::Configuration.load_plugins },
-      :rich_data => false
+      :rich_data => proc { Puppet.lookup(:current_environment).rich_data? }
```

Comment:
This proc will be called within a `Puppet.lookup` which means we're doing recursive `Puppet.lookup` calls, which is the thing I'm _most_ concerned about with this approach.

The other thing is that it will blow up if `current_environment` is not set. However, the current environment is set in the bootstrap_context, which should always be set before the base_context.

Reply:
I think the least invasive change would be for whatever calls `Catalog#to_data_hash` to push `rich_data: true` on the context:

```ruby
Puppet.override(rich_data: true) do
  pp catalog.to_data_hash
end
```

That's essentially what the `rich_data_json` formatter is doing: https://github.com/puppetlabs/puppet/blob/b7f89c426f97fb03a2de884e5372d0ab0edf5c06/lib/puppet/network/formats.rb#L275-L279

However, it perpetuates the caller having to know to do that, for example catalog-diff has a similar issue https://github.com/voxpupuli/puppet-catalog_diff/issues/105#issuecomment-1736355145

Since `rich_data` can be enabled per-environment in `environment.conf`, your changes makes sense. But since it's broken now, do we want to support that case? I'm thinking not, and just use the global setting:

```suggestion
      :rich_data => Puppet[:rich_data]
```

We're already resolving settings in the `base_context` method, so there should be any bootstrap/ordering issues nor should we need to use a `proc`.

Reply:
So there's a "settings" object passed into the base_context method and in my functional testing with a PE install using it worked. However, it did not work in tests, so I went ahead and used the `Puppet[]` access as you mentioned.

I've added a test and updated tests that now fail. As far as I can tell all but one were places where we were depending on the rich_data being false by default for our test assumptions.

The one weird place is that rich_data doesn't seem like it's valid hiera data (or at least unexpected) and so the resource application test failed. To fix that I updated the resource application itself to set rich_data to false. Not sure if that should be pushed down into the call to `to_hiera_hash` or not?